x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -492,6 +507,32 @@ synchronized void handleFailure(Exception e) {
         }
     }
 
+    /**
+     * Cleanup old checkpoints
+     *
+     * @param listener listener to call after done
+     */
+    private void cleanupOldCheckpoints(ActionListener<Void> listener) {
+        long now = getTime();
+        long checkpointLowerBound = context.getCheckpoint() - NUMBER_OF_CHECKPOINTS_TO_KEEP;
+        long lowerBoundEpochMs = now - RETENTION_OF_CHECKPOINTS_MS;
+
+        if (checkpointLowerBound > 0 && lowerBoundEpochMs > 0) {
+            transformsConfigManager.deleteOldCheckpoints(
+                transformConfig.getId(),
+                checkpointLowerBound,
+                lowerBoundEpochMs,
+                ActionListener.wrap(deletes -> {
+                    logger.debug("[{}] deleted [{}] outdated checkpoints", getJobId(), deletes);
+                    listener.onResponse(null);
+                }, listener::onFailure)
```

Comment:
I am not convinced that this should cause `onFinish` to fail. 

We should potentially audit the failure (definitely log it), but failing the `onFinish` flow due to failing to delete old data seems bad to me.